google_guest_agent/main.go

```diff
@@ -183,35 +183,52 @@ func run(ctx context.Context) {
 	logger.Infof("GCE Agent Stopped")
 }
 
-func runCmdOutput(cmd *exec.Cmd) (string, error) {
-	output, err := cmd.StdoutPipe()
-	if err != nil {
-		return "", err
-	}
-	if err := cmd.Start(); err != nil {
-		return "", err
-	}
-	ret, err := ioutil.ReadAll(output)
-	if err != nil {
-		return "", err
-	}
-	if err := cmd.Wait(); err != nil {
-		return "", err
-	}
+type execResult struct {
+	// Return code. Set to -1 if we failed to run the command.
+	code int
+	// Stderr or err.Error if we failed to run the command.
+	err string
+	// Stdout or "" if we failed to run the command.
+	out string
+}
 
-	return string(ret), nil
+func (e execResult) Error() string {
+	return e.err
+}
+
+func (e execResult) ExitCode() int {
+	return e.code
+}
+
+func (e execResult) Stdout() string {
+	return e.out
+}
+
+func (e execResult) Stderr() string {
+	return e.err
 }
 
-// runCmd is exec.Cmd.Run() with a flattened error return.
 func runCmd(cmd *exec.Cmd) error {
+	res := runCmdOutput(cmd)
+	if res.ExitCode() != 0 {
+		return res
+	}
+	return nil
+}
+
+func runCmdOutput(cmd *exec.Cmd) *execResult {
+	var stdout, stderr bytes.Buffer
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+
 	err := cmd.Run()
 	if err != nil {
 		if ee, ok := err.(*exec.ExitError); ok {
-			return fmt.Errorf(string(ee.Stderr))
+			return &execResult{code: ee.ExitCode(), err: stderr.String()}
```

Comment:
For future proofing I would add stdout here as well, just because its non-zero doesn't mean there stderr

Reply:
Actually, just always return &execResult{code: ee.ExitCode(), out: stdout.String(), err: stderr.String()}, it'll make it cleaner.

Reply:
i don't know what you mean by 'always'. this can only be done if the error was type `*exec.ExitError`, otherwise there is no ee. so you mean only to change this one line, right?